### PR TITLE
Copy honors `citus.max_shared_pool_size`

### LINKED ISSUE #3894: COPY should honor `citus.max_shared_pool_size`
`COPY` command is already capable of multiplexing `COPY` commands of multiple shards over a single connection with https://github.com/citusdata/citus/pull/2780.

We should expand it such that even concurrent COPY commands shouldn't use more than `citus.max_shared_pool_size` connections to any worker node from the coordinator across sessions.

Related to #1047, #3496 and #710

### PR DESCRIPTION
**Pull request by onderkalaci**
DESCRIPTION: Improves COPY by honoring max_shared_pool_size config

Simply rely on the existing infrastructure. 

TODO:
- [x] One thing that I'm not very satisfied here is that we may pass `CONNECTION_PER_PLACEMENT` and `OPTIONAL_CONNECTION` flags at the same time. That feels a little counter intuitive, any suggestion to improve  that is  welcome. 
- [x] Do we need to do anything specific for `append/range/reference` tables?
- [x] Are  there any edge cases with sequential mode?

Fixes #3894

### CODE REVIEW

**Comment by hanefi** [comment]
My comments on the outstanding TODO items:

- One thing that I'm not very satisfied here is that we may pass `CONNECTION_PER_PLACEMENT` and `OPTIONAL_CONNECTION` flags at the same time. That feels a little counter intuitive, any suggestion to improve  that is  welcome.
  - If both of these flags are set, we still try to honor `max_shared_pool_size`. If we are not successful in creating a new optional connection, we wait until a new connection can be established. In this regard, I can say that `OPTIONAL_CONNECTION` takes precedence over `CONNECTION_PER_PLACEMENT`. If we have enough capacity in shared connection pool however, we still try to create connections per co-located placements as we did in the past.

- Do we need to do anything specific for `append/range/reference` tables?
  - I tried to test things manually, and did not come up with any edge cases yet. I may need some guidance in here, and it can be possible that I am missing something

- Are  there any edge cases with sequential mode?
  - Similar to the item above, I tried to test different scenarios and I could not find a way to break this PR. I tried setting `multi_shard_modify_mode` to sequential, and copies progressed as the way I expect them.

File: src/backend/distributed/commands/multi_copy.c [hunk 1]
```diff
@@ -3578,6 +3600,18 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
...
+	{
+		/*
+		 * The connection manager throttled any new connections, so pick an existing
+		 * connection with least utilization.
+		 */
+		connection =
+			GetLeastUtilisedCopyConnection(copyConnectionStateList, nodeName, nodePort);
+		Assert(connection != NULL);
+
+		return connection;
```
Comment (hanefi) [inline src/backend/distributed/commands/multi_copy.c:3613]: I wonder if it's ok to return an existing query without checking if the connection is ok, as we do in the following lines:

```c
if (PQstatus(connection->pgConn) != CONNECTION_OK)
```

**Comment by pykello** [comment]
>     * Do we need to do anything specific for `append/range/reference` tables?
>       
>       * I tried to test things manually, and did not come up with any edge cases yet. I may need some guidance in here, and it can be possible that I am missing something

For append tables, we will use `CopyToNewShards()`, which copies to shards sequentially and has `citus.shard_count` connections open at any time. I think it doesn't go through the code path changed in this PR.

Range tables and reference tables use the same code path as hash partitioned table, and `FindShardInterval()` abstracts away the differences. So no need for any specific code path.

File: src/test/regress/sql/shared_connection_stats.sql [hunk 2]
```diff
@@ -190,6 +190,113 @@ BEGIN;
 		hostname, port;
 COMMIT;
 
+BEGIN;
```
Comment (pykello) [inline src/test/regress/sql/shared_connection_stats.sql:193]: 1. Maybe also add similar tests for append distributed, range distributed, and reference tables, and verify that our understanding of how they should behave is correct. To create range distributed table shards, you can use `create_range_partitioned_shards`.
2. Can you try to see if we can use `\COPY ... FROM PROGRAM 'seq 32'` to keep the test shorter?

Comment (marcocitus) [reply src/test/regress/sql/shared_connection_stats.sql:193]: also insert..select tests that uses the COPY logic underneath (especially with RETURNING/ON CONFLICT, because the logic is a bit different).

File: src/backend/distributed/master/master_stage_protocol.c [hunk 3]
```diff
@@ -862,6 +870,12 @@ WorkerShardStats(ShardPlacement *placement, Oid relationId, const char *shardNam
 	MultiConnection *connection = GetPlacementConnection(connectionFlags, placement,
 														 NULL);
 
+	/*
+	 * This code-path doesn't support optional connections, so we don't expect
+	 * NULL connections.
+	 */
+	Assert(connection != NULL && (connectionFlags & OPTIONAL_CONNECTION) == 0);
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/master/master_stage_protocol.c:877]: these asserts were added mostly because of the Microsoft static code analyzers not to complain. 
(I was not sure whether to add, but added to prevent those tools fails because on 9.3 I remember we had to add those asserts https://github.com/citusdata/citus/pull/3805/files)

File: src/backend/distributed/commands/multi_copy.c [hunk 4]
```diff
@@ -3465,17 +3557,50 @@ CopyGetPlacementConnection(ShardPlacement *placement, bool stopOnFailure)
 	}
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c:3557]: is it ok that we're not calling `MarkRemoteTransactionCritical` here?

File: src/backend/distributed/commands/multi_copy.c [hunk 5]
```diff
@@ -85,6 +85,7 @@
 #include "distributed/remote_transaction.h"
 #include "distributed/resource_lock.h"
 #include "distributed/shard_pruning.h"
+#include "distributed/shared_connection_stats.h"
```
Comment (onderkalaci) [inline src/backend/distributed/commands/multi_copy.c:88]: @hanefi can you please rebase this branch on master?

File: src/backend/distributed/commands/multi_copy.c (hunk 1 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/multi_copy.c:3602]: As we early return in many cases now, we should properly use the following code block when necessary:
```C
	MarkRemoteTransactionCritical(connection);

	if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
	{
		ClaimConnectionExclusively(connection);
	}
```

We should call this if the connection is NOT returned from `GetLeastUtilisedCopyConnection`, meaning that we're using the connection for this COPY command for the first time.

File: src/backend/distributed/commands/multi_copy.c (hunk 4 above)
Comment (hanefi) [reply src/backend/distributed/commands/multi_copy.c]: I want to use `MarkRemoteTransactionCritical` along with `ClaimConnectionExclusively`.

Running some more tests for making sure consecutive COPY queries do not try to claim a connection that is not unclaimed by the previous one

File: src/test/regress/sql/shared_connection_stats.sql (hunk 2 above)
Comment (marcocitus) [reply src/test/regress/sql/shared_connection_stats.sql:193]: bump

File: src/backend/distributed/commands/multi_copy.c [hunk 6]
```diff
@@ -3554,30 +3567,53 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
...
+		connection = GetLeastUtilisedCopyConnection(copyConnectionStateList,
+													nodeName,
+													nodePort);
+
+		/*
+		 * Errors are supposed to cause immediate aborts (i.e. we don't
+		 * want to/can't invalidate placements), mark the connection as
+		 * critical so later errors cause failures.
+		 */
+		MarkRemoteTransactionCritical(connection);
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c:3613]: Given that it's in copyConnectionStateList, we should have already done this.

File: src/backend/distributed/commands/multi_copy.c [hunk 7]
```diff
@@ -3554,30 +3567,53 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
...
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c:3583]: This should be below `GetConnectionIfPlacementAccessedInXact`, not here.

File: src/backend/distributed/commands/multi_copy.c (hunk 1 above)
Comment (hanefi) [reply src/backend/distributed/commands/multi_copy.c:3615]: I run into issues when I tried to mark the connections that were used to access placements as exclusive.

You can see my attempt in this commit: 87f185ba102eb016109e2b6cfc0d87bf86b9c604

Here is one of the places where copy started to fail, after I marked it as exclusive:
https://github.com/citusdata/citus/blob/87f185ba102eb016109e2b6cfc0d87bf86b9c604/src/test/regress/expected/multi_modifying_xacts.out#L221-L227

File: src/backend/distributed/connection/placement_connection.c [hunk 8]
```diff
@@ -209,6 +210,14 @@ GetPlacementConnection(uint32 flags, ShardPlacement *placement, const char *user
 {
 	MultiConnection *connection = StartPlacementConnection(flags, placement, userName);
 
+	if (connection == NULL)
+	{
+		/* connection can only be NULL for optional connections */
+		Assert((flags & OPTIONAL_CONNECTION));
```
[the remark this answers is not in the document]
Comment (hanefi) [reply src/backend/distributed/connection/placement_connection.c:214]: `CopyGetPlacementConnection` can set this flag and call `GetPlacementConnection` in lines:

```c
	/*
	 * Enforce the requirements for adaptive connection connection
	 * management (a.k.a., throttle connections if citus.max_shared_pool_size
	 * reached)
	 */
	int adaptiveConnectionManagementFlag =
		AdaptiveConnectionManagementFlag(list_length(copyConnectionStateList));
	connectionFlags |= adaptiveConnectionManagementFlag;

	connection = GetPlacementConnection(connectionFlags, placement, nodeUser);
```

I may as well move the assertion to the caller, if that helps

File: src/backend/distributed/commands/multi_copy.c (hunk 4 above)
Comment (marcocitus) [reply src/backend/distributed/commands/multi_copy.c:3586]: I think the main effect of not calling ClaimConnectionExclusively would be that, if a connection already wrote to a shard and the shard is the first shard that the COPY writes to, we might start using that connection for multiple shard , instead of opening separate connections per shard.

File: src/backend/distributed/commands/multi_copy.c (hunk 1 above)
Comment (marcocitus) [reply src/backend/distributed/commands/multi_copy.c:3630]: hmz, something is a little off here.

The error message is right and the bug is that we should not be claiming connections that still need to be used for other placements, but then is it ever ok to claim connections on the COPY logic? (i.e. do we have a bug even without the extra call?)

In the adaptive executor, we first assign tasks to connections, and only them claim the connections exclusively. We know that after this, any subsequent task does not have an assigned connection, so we never reach a point where we need to use a specific connection when that connection is claimed exclusively.

Comment (marcocitus) [reply src/backend/distributed/commands/multi_copy.c:3630]: I guess CONNECTION_PER_PLACEMENT saved us in the past by ensuring we only claim clean connections, so we never ran into a situation where GetConnectionIfPlacementAccessedInXact errors out later. Maybe revert that commit?

Comment (marcocitus) [reply src/backend/distributed/commands/multi_copy.c:3630]: I guess what we would really want is some variant of "claimed exclusively" which instead of "return this connection to no one" means "return this connection only to me".

File: src/test/regress/sql/shared_connection_stats.sql (hunk 2 above)
Comment (hanefi) [reply src/test/regress/sql/shared_connection_stats.sql:193]: Added some more tests
- [x] `INSERT INTO SELECT ... RETURNING *`
- [x] range distribution
- [x] reference tables
- [ ] append distribution

File: src/backend/distributed/connection/placement_connection.c [hunk 9]
```diff
@@ -1175,6 +1115,19 @@ InitPlacementConnectionManagement(void)
 }
 
 
+/*
+ * UseConnectionPerPlacement returns whether we should use as separate connection
+ * per placement even if another connection is idle. We mostly use this in testing
+ * scenarios.
+ */
+bool
+UseConnectionPerPlacement(void)
```
Comment (onurctirtir) [inline src/backend/distributed/connection/placement_connection.c:1124]: maybe `CanUseConnectionPerPlacement` or `ShouldUseConnectionPerPlacement` as it does not change the state

File: src/backend/distributed/master/master_stage_protocol.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanefi) [reply src/backend/distributed/master/master_stage_protocol.c]: Left the assertion as `Assert(connection != NULL)`

File: src/backend/distributed/commands/multi_copy.c [hunk 10]
```diff
@@ -3262,6 +3324,36 @@ ConnectionStateList(HTAB *connectionStateHash)
 }
 
 
+/*
+ * ConnectionStateListToNode returns all CopyConnectionState structures in
+ * the given hash.
+ */
+static List *
+ConnectionStateListToNode(HTAB *connectionStateHash, char *hostname, int port)
```
[the remark this answers is not in the document]
Comment (hanefi) [reply src/backend/distributed/commands/multi_copy.c]: These are nice practices that I occasionally forget. Thanks for reminding me .

File: src/backend/distributed/commands/multi_copy.c (hunk 6 above)
Comment (hanefi) [reply src/backend/distributed/commands/multi_copy.c]: I removed the line

File: src/backend/distributed/connection/placement_connection.c (hunk 9 above)
Comment (hanefi) [reply src/backend/distributed/connection/placement_connection.c]: We decided to rename `CONNECTION_PER_PLACEMENT` to `REQUIRE_CLEAN_CONNECTION`
Thanks for reminding me that there are other places where I need to rename.

How does eda7d8e0d2b6281b68223f3b78d4f9c79a53be0d look? I renamed this function to `ShouldUseCleanConnection`

Comment (onurctirtir) [reply src/backend/distributed/connection/placement_connection.c]: looks good as the enum implies

**Comment by hanefi** [comment]
I will squash commits, and resolve enterprise merge problem. I will waiting until  I get another round of reviews/approvals

File: src/backend/distributed/connection/placement_connection.c [hunk 11]
```diff
@@ -1175,6 +1177,18 @@ InitPlacementConnectionManagement(void)
 }
 
 
+/*
+ * ShouldUseCleanConnection returns whether we should force a new connection if
+ * existing connections accessed non-co-located placements on the workers.
+ */
+bool
+ShouldUseCleanConnection(void)
```
Comment (marcocitus) [inline src/backend/distributed/connection/placement_connection.c:1185]: Can you revert this name? It seems unrelated to the flag.

File: src/backend/distributed/commands/multi_copy.c [hunk 12]
```diff
@@ -3596,10 +3624,22 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
 	if (placement->partitionMethod == DISTRIBUTE_BY_HASH &&
 		MultiShardConnectionType != SEQUENTIAL_CONNECTION)
 	{
-		connectionFlags |= CONNECTION_PER_PLACEMENT;
+		connectionFlags |= REQUIRE_CLEAN_CONNECTION;
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c:3627]: Maybe expand the comment to capture our current knowledge:

Claiming the connection exclusively (done below) would also have the effect of opening multiple connections, but claiming the connection exclusively prevents GetConnectionIfPlacementAccessedInXact from returning the connection if it is needed for a different shard placement. By setting the REQUIRE_CLEAN_CONNECTION flag we are guaranteed to get connection that will not be returned by GetConnectionIfPlacementAccessedInXact for the remainder of the COPY, hence it safe to claim the connection exclusively. Claiming a connection exclusively prevents it from being used in other  distributed queries that happen during the COPY (e.g. if the copy logic calls a function to calculate a default value, and the function does a distributed query).

File: src/backend/distributed/connection/placement_connection.c (hunk 11 above)
Comment (hanefi) [reply src/backend/distributed/connection/placement_connection.c]: Done. I left it as the original `UseConnectionPerPlacement()`.

I thought that once we renamed the flag `s/CONNECTION_PER_PLACEMENT/REQUIRE_CLEAN_CONNECTION/`, it makes sense to replace occurences of `s/ConnectionPerPlacement/CleanConnection/`.

I'm ok with either way

File: src/backend/distributed/commands/multi_copy.c (hunk 4 above)
Comment (marcocitus) [reply src/backend/distributed/commands/multi_copy.c:3586]: There's a more subtle effect that is worth documenting in a comment: 

TODO: We do not claim the connection exclusively, because may need to be returned by GetConnectionIfPlacementAccessedInXact multiple times and claiming exclusively prevents that. That means we are not guarded against distributed queries that happen during the COPY (e.g. a distributed query in a function that is called to compute a default value). The proper solution would be to claim all connections which already modified shard placements at the start of the COPY, regardless of whether rows are sent to those shard placements during the COPY.

File: src/backend/distributed/commands/multi_copy.c [hunk 13]
```diff
@@ -3576,18 +3596,26 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
...
-			GetLeastUtilisedCopyConnection(copyConnectionStateList, nodeName, nodePort);
-
 		/*
 		 * If we've already reached the executor pool size, there should be at
 		 * least one connection to any given node.
 		 */
-		Assert(connection != NULL);
+		connection = GetLeastUtilisedCopyConnection(copyConnectionStateList,
+													nodeName,
+													nodePort);
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c:3606]: Maybe add: We don't need to mark the connection as critical, since it was already returned by this function before.

File: src/backend/distributed/connection/placement_connection.c (hunk 11 above)
Comment (marcocitus) [reply src/backend/distributed/connection/placement_connection.c]: This mostly just shows how poorly the CONNECTION_PER_PLACEMENT flag was named. These two concepts are unrelated, but had the same name.

File: src/backend/distributed/commands/multi_copy.c [hunk 14]
```diff
@@ -3596,10 +3627,39 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
...
-		connectionFlags |= CONNECTION_PER_PLACEMENT;
+		/*
+		 * Claiming the connection exclusively (done below) would also have the
+		 * effect of opening multiple connections, but claiming the connection
+		 * exclusively prevents GetConnectionIfPlacementAccessedInXact from returning
+		 * the connection if it is needed for a different shard placement.
+		 *
+		 * By setting the REQUIRE_CLEAN_CONNECTION flag we are guaranteed to get
+		 * connection that will not be returned by GetConnectionIfPlacementAccessedInXact
+		 * for the remainder of the COPY, hence it safe to claim the connection
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/commands/multi_copy.c:3638]: it is safe ...

File: src/backend/distributed/connection/shared_connection_stats.c [hunk 15]
```diff
@@ -587,6 +589,86 @@ SharedConnectionStatsShmemInit(void)
...
+		/*
+		 * User wants one connection per placement, so no throttling is desired
+		 * and we do not set any flags.
+		 *
+		 * The primary reason for this is that allowing multiple backends to use
+		 * connection per placement could lead to unresolved self deadlocks. In other
+		 * words, each backend may stuck waiting for other backends to get a slot
+		 * in the shared connection counters.
+		 */
+		return 0;
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/connection/shared_connection_stats.c:613]: What does 0 mean? Using a macro, or something to remove the mapping could be useful

**Comment by marcocitus** [comment]
We should make sure that we raise the shared connection counter for all workers to 1 before continuing with the COPY (and grabbing optional connections).

**Comment by hanefi** [comment]
I want to share some offline discussions for documentation purposes:

> If the database receive concurrent SELECT and COPY queries, we can run into deadlocks as different queries can use essential connections, and are unable to create optional connections to the remaining placements.
> 
> It appears that COPY tries to get connections to placements in the order decided by the tuples that we receive via `CitusTupleDestReceiver`. We used to be able to create new connections if there was no cached connection. However now COPY can reserve an essential connection to a placement, only to get stuck waiting for connecting to other placements.
> 
> I could come up with a single solution: Somehow try to create connections in the same order that multi-shard SELECT commands would. As we do not know all the shards that will be hit by a COPY beforehand, this is not a trivial matter.

Here is an idea on how to solve this possible connection starvation issue: 

A COPY operation should have the right to establish connections to all placements   during `CitusCopyDestReceiverStartup`. One way to accomplish this is raising the connection counters, but defer the actual connection open to a later time.

**Comment by marcocitus** [comment]
Superseded by #4034
